odl/space/fspace.py

```diff
@@ -1141,26 +931,334 @@ def __init__(self, fspace, fcall):
+                x = x[0][None, ...]
+        elif is_valid_input_array(x, ndim):
+            x = np.asarray(x)
+            scalar_in = False
+            scalar_out_shape = out_shape_from_array(x)
+            scalar_out = False
+        elif x in self.domain:
+            x = np.atleast_2d(x).T  # make a (d, 1) array
+            scalar_in = True
+            scalar_out_shape = (1,)
+            scalar_out = (out is None and not self.space.tensor_valued)
+        else:
+            # Unknown input
+            txt_1d = ' or (n,)' if ndim == 1 else ''
+            raise TypeError('argument {!r} not a valid function '
+                            'input. Expected an element of the domain '
+                            '{domain}, an array-like with shape '
+                            '({domain.ndim}, n){} or a length-{domain.ndim} '
+                            'meshgrid tuple.'
+                            ''.format(x, txt_1d, domain=self.domain))
+
+        # Check bounds if specified
+        if bounds_check:
+            if not self.domain.contains_all(x):
+                raise ValueError('input contains points outside '
+                                 'the domain {}'.format(self.domain))
+
+        if scalar_in:
+            out_shape = self.out_shape
+        else:
+            out_shape = self.out_shape + scalar_out_shape
+
+        # Call the function and check out shape, before or after
+        if out is None:
+            if ndim == 1:
+                try:
+                    out = self._call(x, **kwargs)
+                except (TypeError, IndexError):
+                    # TypeError is raised if a meshgrid was used but the
+                    # function expected an array (1d only). In this case we try
+                    # again with the first meshgrid vector.
+                    # IndexError is raised in expressions like x[x > 0] since
+                    # "x > 0" evaluates to 'True', i.e. 1, and that index is
+                    # out of range for a meshgrid tuple of length 1 :-). To get
+                    # the real errors with indexing, we check again for the
+                    # same scenario (scalar output when not valid) as in the
+                    # first case.
+                    out = self._call(x[0], **kwargs)
+
+            else:
+                # Here we don't catch exceptions since they are likely true
+                # errors
+                out = self._call(x, **kwargs)
+
+            if isinstance(out, np.ndarray) or np.isscalar(out):
+                # Cast to proper dtype if needed, also convert to array if out
+                # is a scalar.
+                out = np.asarray(out, dtype=self.space.scalar_out_dtype)
+                if scalar_in:
+                    out = np.squeeze(out)
+                elif ndim == 1 and out.shape == (1,) + out_shape:
+                    out = out.reshape(out_shape)
+
+                if out_shape not in ((), (1,)) and out.shape != out_shape:
```

Comment:
Actually only the `()` case should be filtered out here, since we want
```python
>>> fspace = odl.FunctionSpace(odl.IntervalProd(0, 1), out_dtype=(float, (1,)))
>>> func = fspace.element(lambda x: x)
>>> func(0)
array([ 0.])
```
Fixing